scripts/build/libc/picolibc.sh

```diff
@@ -0,0 +1,116 @@
+# This file adds functions to build the Picolibc library
+# Copyright © 2020 Keith Packard
+# Licensed under the GPL v2 or later. See COPYING in the root of this package
+#
+# Edited by Keith Packard <keithp@keithp.com>
+#
+
+picolibc_start_files()
+{
+    CT_DoStep INFO "Installing C library headers & start files"
+    CT_DoExecLog ALL cp -a "${CT_SRC_DIR}/picolibc/newlib/libc/include/." \
+    "${CT_HEADERS_DIR}"
+    if [ "${CT_ARCH_XTENSA}" = "y" ]; then
+        CT_DoLog EXTRA "Installing Xtensa headers"
+        CT_DoExecLog ALL cp -r "${CT_SRC_DIR}/picolibc/newlib/libc/sys/xtensa/include/."   \
+                               "${CT_HEADERS_DIR}"
+    fi
+    CT_EndStep
+}
+
+picolibc_main()
+{
+    local -a picolibc_opts
+    local cflags_for_target
+
+    CT_DoStep INFO "Installing C library"
+
+    CT_mkdir_pushd "${CT_BUILD_DIR}/build-libc"
+
+    CT_DoLog EXTRA "Configuring C library"
+
+    # Multilib is the default, so if it is not enabled, disable it.
+    if [ "${CT_MULTILIB}" != "y" ]; then
+        picolibc_opts+=("-Dmultilib=false")
+    fi
+
+    yn_args="IO_C99FMT:io-c99-formats
+IO_LL:io-long-long
+REGISTER_FINI:newlib-register-fini
+NANO_MALLOC:newlib-nano-malloc
+ATEXIT_DYNAMIC_ALLOC:newlib-atexit-dynamic-alloc
+GLOBAL_ATEXIT:newlib-global-atexit
+LITE_EXIT:lite-exit
+MULTITHREAD:newlib-multithread
+RETARGETABLE_LOCKING:newlib-retargetable-locking
+    "
+
+    for ynarg in $yn_args; do
+        var="CT_LIBC_PICOLIBC_${ynarg%:*}"
+        eval var=\$${var}
+        argument=${ynarg#*:}
+
+
+        if [ "${var}" = "y" ]; then
+            picolibc_opts+=( "-D$argument=true" )
+        else
+            picolibc_opts+=( "-D$argument=false" )
+        fi
+    done
+
+    [ "${CT_USE_SYSROOT}" = "y" ] && \
+	picolibc_opts+=( "-Dsysroot-install=true" )
+
+    [ "${CT_LIBC_PICOLIBC_EXTRA_SECTIONS}" = "y" ] && \
+        CT_LIBC_PICOLIBC_TARGET_CFLAGS="${CT_LIBC_PICOLIBC_TARGET_CFLAGS} -ffunction-sections -fdata-sections"
+
+    [ "${CT_LIBC_PICOLIBC_LTO}" = "y" ] && \
+        CT_LIBC_PICOLIBC_TARGET_CFLAGS="${CT_LIBC_PICOLIBC_TARGET_CFLAGS} -flto"
+
+    cflags_for_target="${CT_ALL_TARGET_CFLAGS} ${CT_LIBC_PICOLIBC_TARGET_CFLAGS}"
+
+    # Note: picolibc handles the build/host/target a little bit differently
+    # than one would expect:
+    #   build  : not used
+    #   host   : the machine building picolibc
+    #   target : the machine picolibc runs on
+	meson_cflags=""
+	for cflag in ${cflags_for_target}; do
+		meson_cflags="${meson_cflags} '${cflag}',"
+	done
+	cat << EOF > picolibc-cross.txt
```

Comment:
I see this adds a new host dependency on meson and ninja. Which is fine but we probably should back that up with some kind of configure time check that only allows picolibc to be selected if these tools are present. We've got such an option for lzip, wget and curl.

Reply:
ok, I'll work on adding those checks. I'm also working on getting picolibc building on Windows using mingw and msys tools. Is that something that would be useful here, or is this just for building tooling for real computers?

Reply:
There currently isn't much real support for anything other than Linux.

With the right tools (i.e. GNU sed, awk, stat) various BSDs are supportable. macOS support has somewhat atrophied due to lack of use. I'd be interested in hearing from anyone who has tried building crosstool-ng under WSL, it might just work. Cygwin/msys2/mingw will probably be limited by the file system adaptation as ct-ng is a bit picky about being able to set permissions.

Reply:
ok, checks added for meson and ninja. I've done some limited testing and things 'seem' to work. It would be cool to set up CI here on github and run the picolibc test suite using the built tooling.

Reply:
Yeah CI is a bit of a problem for ct-ng with the sheer number of permutations and the time it takes to build a compiler. travis-ci times out at 50 minutes for a build job so we were plagued by failures due to timeout, I managed to get shadow banned on gitlab-ci while testing out using that (it allows timeout configuration but has a per-user allocation of CPU time, It think I'm still about 9000 minutes in debt).

Azure looks promising and somewhat friendly to FLOSS projects (as long as I don't try my world build on it).

Reply:
I'm running 200 separate configurations of picolibc on github these days and it seems to be working; it takes about 90 minutes to run, depend on how many github runs in parallel. Not using autotools is probably saving me though; I'm compiling somewhere around 1.2 million files.

Reply:
Hmm OK. I obviously haven't been keeping up with github actions.

Reply:
Yeah, once I had someone show me how to avoid depending on an external dockerhub image, I've been pretty happy with how easy it is to use them for CI. Picolibc now runs tests on every push and PR. https://github.com/picolibc/picolibc/actions